### ISSUE TITLE
Chrome Web Store: "Removal notification for uBlock Origin"

### ISSUE THREAD

**Issue by gorhill** [gorhill/uBlock#880]
Note: _as of writing_ the extension is still in the store -- despite the title of the email I received.

Sigh.

I received the following email regarding my submission of 1.3.2 to the Chrome store:

> Subject: Chrome Web Store: Removal notification for uBlock Origin
> From: chromewebstore-dev-support@google.com
> Cc: chromewebstore-dev-support@google.com
> 
> Dear Developer,
> 
> Your Google Chrome item, "uBlock Origin," with ID: cjpalhdlnbpafiamejdnhcphjbkeiagm did not comply with the following section of our Program policies:
> 
> "Where possible, make as much of your code visible in the package as you can. If some of your app's logic is hidden and it appears to be suspicious, we may remove it."
> 
> If you'd like to re-submit your item, please make appropriate changes to the item so that it complies with our policies, then re-publish it in your developer dashboard. Please reply to this email for issues regarding this item removal.
> 
> *Please keep in mind that your re-submitted item will not be immediately published live in the store. All re-submitted items undergo a strict compliance review and will be re-published if the item passes review.
> 
> *Important Note
> Repeated or egregious violations in the store may result in your developer account being banned from the store. This may also result in the suspension of related Google services associated with your Google account. All re-submitted items will continue to be subject to Chrome Web Store policies and terms of service.
> 
> Thank you for your cooperation,
> Google Chrome Web Store team
> 
> ---------------------------
> Developer Terms of Service:
> https://developers.google.com/chrome/web-store/terms
> 
> Program Policies:
> https://developers.google.com/chrome/web-store/program_policies
> 
> Branding Guidelines:
> https://developers.google.com/chrome/web-store/branding 

This is the [diff between uBO 1.2.0 and 1.3.2](https://github.com/gorhill/uBlock/compare/1.2.0...1.3.2). Many of the commits affect FF branch only.

Nothing out of the ordinary with 1.3.2: bug fixes and the only "new" feature (FF already had this implemented) is to read an entry from managed storage so that admins can enforce their own uBO settings ([commit](https://github.com/gorhill/uBlock/commit/892913d185d8aa2eaa7ab10aa812a5fa3814f2ea)).

So essentially, the policy which is pointed out...

> "Where possible, make as much of your code visible in the package as you can. If some of your app's logic is hidden and it appears to be suspicious, we may remove it."

... doesn't tell me anything specific regarding the issue. uBO's code base is rather large, so I don't see how I can ever resolve this issue without specifics. There is no obfuscated code in uBO, it's rather well commented throughout, and I have no problem describing whatever piece of code does what. But having to guess what is the issue here given the code base seems like mission impossible.

**Comment by gorhill** [comment]
Mainly this amounts to: "There is _somewhere_ one or more pieces of code I don't understand, but I won't tell you what it is I don't understand. Your challenge is to find what I am talking about and modify it so that in my next review I will _maybe_ understand it."

**Comment by Wingie** [comment]
but does this mean the old version is still okay to be on the store?

**Comment by gorhill** [comment]
> but does this mean the old version is still okay to be on the store?

I don't think so, the email is titled "Removal notification". 1.3.2 was not in the Chrome store, it was merely submitted for publication, so it did not need to be removed -- just rejected.

Even assuming if it would still be ok, that pretty much mean it is dead, since from now on I can't update it anymore (there were good bug fixes in 1.3.2). There is no way I will ever figure what is the specific issue given how large uBO's code base is.

**Comment by kangman** [comment]
haven't they heard of github?

**Comment by Wingie** [comment]
maybe they can only understand it if it is on google code...

Seriously though, this reminds me of [streamus](https://www.reddit.com/r/streamus/comments/3dwl3p/streamus_download_and_installation_newest_version/) and google (youtube)'s battle a few months back. The streamus developer struggled to bring his extensions upto google's standards and they just kept making absurd reasons for denial until he finally gave in. They made up a lot of "technical" reasons in his case too. you can go through the subreddit for more details. And of course, the bottom line is that he was cutting into their ad revenue by not being able to play youtube ads in his extensions.. 

I would advise you to look into an alternative delivery mechanism for the extension. Google just can't be trusted anymore..

**Comment by rmndk** [comment]
> Please reply to this email for issues regarding this item removal.

@gorhill  Did you reply yet?

**Comment by gorhill** [comment]
> @gorhill Did you reply yet?

My [previous experience](https://github.com/gorhill/uBlock/issues/50#issuecomment-93021229) with the people behind "Chrome store" emails is that they never provide specifics --

**Comment by killswitch** [comment]
> 2,160,627 users

They're not very smart are they? "I don't understand code, so lets take down a wildly popular extension for no reason whatsoever."

**Comment by sowbug** [comment]
Hi everyone, I've forwarded this issue to Chrome Web Store's abuse team. Sorry for the hassle; we'll get it resolved as soon as we can.

**Comment by Sp1l** [comment]
["Chrome Web Store is Google's online store"](https://en.wikipedia.org/wiki/Chrome_Web_Store)
Hold on there!!! Wasn't Google also the owner of some Advertisement network???
["Google AdWords is an online advertising service"](https://en.wikipedia.org/wiki/AdWords)

**Comment by RobinJ1995** [comment]
The reason is at least still a lot more specific that what you'd get from the Play Store... Google needs to get their act together. My guess is that the removal was automated and as such they don't have much information to give. Otherwise, it's just an unwillingness to give more information on their part.
I'd send them an e-mail about it. There's probably an appeal procedure.

**Comment by sebagalante** [comment]
Solution: Firefox

**Comment by TETYYS** [comment]
> Solution: Ignore the problem

**Comment by adamkdean** [comment]
Maybe I'm missing something, but uBlock Origin is still in the CWS and I was just able to install it?

**Comment by svnpenn** [comment]
@adamkdean maybe you should read the full thread before commenting?

**Comment by josephglanville** [comment]
@sowbug awesome. Hopefully they are able to do something before it gets pulled and 1.3.2 is able to go through unhindered. uBO is an essential extension for web performance for many users.

**Comment by BtM909** [comment]
@adamkdean if you read https://github.com/gorhill/uBlock/issues/880#issuecomment-152173152, you'll see that the new(est) version wasn't accepted to be published in the store.

The existing version is still online, but it might be pending to be removed as well.

**Comment by gorhill** [comment]
Received a response from the Chrome store:

> Dear Developer,
> 
> We apologize that the update was rejected due to an snag in the review system. The updated item will be available in the Chrome Web Store within 30 minutes.
> 
> Thank you for your cooperation,
> Google Chrome Web Store team
> 
> ---------------------------
> Developer Terms of Service:
> https://developers.google.com/chrome/web-store/terms
> 
> Program Policies:
> https://developers.google.com/chrome/web-store/program_policies
> 
> Branding Guidelines:
> https://developers.google.com/chrome/web-store/branding

So this makes the issue resolved. (maybe @sowbug [email](https://github.com/gorhill/uBlock/issues/880#issuecomment-152184030) helped?)

Sorry for this, that really got me worried. If this happens again I will at wait a bit more for feedback from the Chrome store before reporting here. Unclear though whether making such issue widely known sooner than later helps with its resolution, or at least a faster one.

**Comment by gorhill** [comment]
The fact that there are no specifics with such warning is still a Chrome store issue though. If this is caused by some kind of automated review, it would be nice to have at least the file(s) + line numbers which triggered the issue, so that devs are not left completely in the dark as to what needs to be worked on, or at least what needs clarification, detailed explanation.

**Comment by pipermerriam** [comment]
> Sorry for this, that really got me worried. If this happens again I will at wait a bit more for feedback from the Chrome store before reporting here. Unclear though whether making such issue widely known sooner than later helps with its resolution, or at least a faster one.

I think that your posting it here was very reasonable and that you shouldn't hesitate to do the same if a similar situation arises again.  It isn't your responsibility to try and decipher whether an email like this is a bug or a real issue.

**Comment by joepie91** [comment]
> I think that your posting it here was very reasonable and that you shouldn't hesitate to do the same if a similar situation arises again. It isn't your responsibility to try and decipher whether an email like this is a bug or a real issue.

+1

**Comment by seanrand** [comment]
> The fact that there are no specifics with such warning is still a Chrome store issue though. If this is caused by some kind of automated review, it would be nice to have at least the files + line numbers which triggered the issue, so that devs are not left completely in the dark as to what need to be worked on, or at least what need clarification, detailed explanation.

Yeah, that's a long-standing issue with the Chrome Web Store (and all the other Google-managed Web/App Stores). I know of several devs that had their products pulled over the years with little to no info and were only able to have them reinstated after generating a sizeable shitstorm on Twitter and social media.

**Comment by ojensen5115** [comment]
> it would be nice to have at least the files + line numbers which triggered the issue, so that devs are not left completely in the dark as to what need to be worked on, or at least what need clarification, detailed explanation.

I agree with this in theory, but in practice I think it's unlikely to happen for the same reasons as spam filters don't respond with the line and word numbers which contributed to an email being marked as spam. It's unfortunate but I think that with the sheer amount of abuse systems like this need to scrub, giving those trying to abuse the system a window into exactly what they need to tweak to evade detection probably just isn't worth not losing the trust of the occasional great developer. You probably couldn't even have extension popularity as a reliable heuristic to not doing this, because a fantastic way to silently push your malware to millions of users is just to buy some popular extension and push an update.

At the end of the day, I suspect it's just a cost/benefit numbers game. I really hope they can find a way to do better in a way that makes sense for them too, though, because situations like these are really frustrating, and can suck the fun out of being a dev :(

**Comment by vjyanand** [comment]
It's not just chrome store, it is same across admob and google play store

**Comment by joepie91** [comment]
> I agree with this in theory, but in practice I think it's unlikely to happen for the same reasons as spam filters don't respond with the line and word numbers which contributed to an email being marked as spam. It's unfortunate but I think that with the sheer amount of abuse systems like this need to scrub, giving those trying to abuse the system a window into exactly what they need to tweak to evade detection probably just isn't worth not losing the trust of the occasional great developer.

If this is a concern, then the system is broken to begin with. Security through obscurity doesn't work, and only benefits the malicious actors in the end (who have the economic incentive to find their way through it).

Perhaps they should fix the review system instead.

**Comment by ojensen5115** [comment]
> If this is a concern, then the system is broken to begin with. Security through obscurity doesn't work

And yet the amount of spam in your inbox is less than 1% of what it was 15 years ago. And SEO is a thing. And so are virus scanners and intrusion detection systems and honeypots.

> Perhaps they should fix the review system instead.

That would be nice. And I think that if anyone discovers a way to accurately solve these problems in a *non*-obscure fashion, and in a way that doesn't trivially invite a denial of service attack by scaling asymmetrically in terms of effort required on behalf of abusers being small as compared to effort required on behalf of verifiers, I think they're likely to become billionaires almost overnight. But in the mean time, we have a messy partial solution to a messy problem. And it's not perfect, and clearly broken in some ways (case in point: this entire thread) -- but you have to admit that it's far from useless.

**Comment by joepie91** [comment]
> And yet the amount of spam in your inbox is less than 1% of what it was 15 years ago. And SEO is a thing. And so are virus scanners and intrusion detection systems and honeypots.

Yet I still frequently see things ending up in Spam by mistake, and it's now practically impossible as a one-man project to get your e-mail through anywhere without paying a third-party service. Spam filtering absolutely doesn't work well.

AV software is an extra layer of defense, but also doesn't solve the problem. It is not a replacement for more secure architecture; it is an addition. It does not analogize well with this situation.

> That would be nice. And if anyone discovers a way to accurately solve these problems in a non-obscure fashion, and in a way that doesn't trivially invite a denial of service attack by scaling asymmetrically in terms of effort required on behalf of abusers being small as compared to effort required on behalf of verifiers, I think they're likely to become billionaires almost overnight.

We already have such solutions. Greylisting-like approaches, similarity checks, new-submitter pre-moderation gates, sandboxing, and so on.

That Google is unwilling to invest in manpower for these solutions, just like they are unwilling to do so for user-facing customer support, is a different discussion entirely.

**Comment by adamkdean** [comment]
@svnpenn I gave it a quick scan but maybe my understanding was skewed by the ambiguous headline. If the update is rejected, it's not the same as "uBlock Origin removed from CWS". No need for snarky comments around here mate.

@BtM909 yeah I got that much, I was just wondering where the "removed from CWS" came from, as if the update is never accepted then how can it ever have been on CWS in order to have been removed? Maybe it's just the ambiguity of everything that through me. Thanks for not being a snarky shit about it though.

**Comment by svnpenn** [comment]
Hey @adamkdean, fuck right off mate!

**Comment by AMKohn** [comment]
@gorhill I'm not here to complain, but I thought you might find it useful to know that this has happened a number of times to me with my extensions (which are pretty sizable).  It apparently literally happens for no good reason. My extensions have been taken down when I haven't made changes in weeks.

I managed to reach someone at the Webstore (in the end I had to email someone who _used_ to work there, who then forwarded my email to someone there and I still didn't get a response for more than a week) who basically said it complies with everything and doesn't look like it matches any filters.  The extensions have also been put back up when I haven't made any changes.

It also doesn't look like there's an end in sight. As best I can tell, Google cut most of the Webstore development team in 2012 (perhaps @sowbug can confirm this?). The copyright on the dashboard hasn't even been updated in almost 3 years.  Google might revamp the system at some point, but it looks like they're happy to let it break down for now.

__TL;DR:__ The Webstore apparently has buggy filters and we have to put up with them or not make Chrome extensions. The removal likely had nothing to do with your code or any change you made.

**Comment by bradvogel** [comment]
This has happened to me a bunch of times. It's completely random. Just reorder some code to make the diff look big and then resubmit. It might take 3-4 tries. They won't pull it from the store though, so you're fine.

**Comment by sowbug** [comment]
@AMKohn good catch on the copyright footer! It's below the fold for active extension developers (including Web Store engineers themselves), so I doubt anyone on the team noticed it. It's fixed -- for 2015, at least -- and will roll out with the next push.

**Comment by AMKohn** [comment]
@sowbug That's great, but it really wasn't a big deal (or the only thing like that, half the links are out of date and get redirected).  I mentioned it because I feel it's an indicator of a general lack of care, at least on the developer side of things. Are there even any development plans for the dashboard?  Maybe something simple like an email notification when someone sends a message via the feedback system (another small thing)?

It seems like Google built Chrome on the backs of developers (unless I've perceived things incorrectly, a major strategy of Google's to grow Chrome has been to offer great development tools and support the latest features, attracting developers who make apps for average users) and now that it's the most popular browser out there, forgotten about the developers that helped to get it there. Without apps and extensions Chrome wouldn't be what it is today.

What I don't understand is why. Chrome apps, via the webstore, are used in more places than ever. Chromebooks run exclusively on them, and even Google Drive addons run through the system. Various other dashboards around Google have had major revamps (the Android system, API console, and apps control panel to name a few) while the webstore dashboard hasn't even had its copyright updated in years.  The last change I remember on the developer side was the addition of the translation system, which felt like it was done quickly to make more money.

Please understand that I'm in no way blaming you (as far as I can tell you don't work on the webstore anyway) or anyone working on the webstore, just management higher up.  I'm sure everyone there is quite busy maintaining the store and doing the best they can, but changes like moving people to other divisions don't happen by accident (please correct me if I'm wrong about the changes in 2012).
